src/Runner.Listener/Runner.cs

```diff
@@ -408,7 +408,7 @@ private async Task<int> RunAsync(RunnerSettings settings, bool runOnce = false)
                                     autoUpdateInProgress = true;
                                     var runnerUpdateMessage = JsonUtility.FromString<AgentRefreshMessage>(message.Body);
                                     var selfUpdater = HostContext.GetService<ISelfUpdater>();
-                                    selfUpdateTask = selfUpdater.SelfUpdate(runnerUpdateMessage, jobDispatcher, !runOnce && HostContext.StartupType != StartupType.Service, HostContext.RunnerShutdownToken);
+                                    selfUpdateTask = selfUpdater.SelfUpdate(runnerUpdateMessage, jobDispatcher, false, HostContext.RunnerShutdownToken);
```

Comment:
If I read this right, we will no longer need to `restartInteractiveRunner` to true as we will take care of it in run.sh|cmd. 

If you agree, I'll clean up this argument and the restart depending on it in `update.sh.template`

@TingluoHuang

Reply:
consider an upgrade from a runner without all your `run.sh`, I think we need to keep some of the legacy code for a while...

Reply:
You're right, it's too early to clean up `update.sh.template`.

Let's say there's `current` and `patched` runner versions and ignore ephemeral and service runners for now.

### This is how it works now:
`current` updating to `current.1`:
 - `SelfUpdate` is called with `restartInteractiveRunner=true`
 - `current` `run.sh` exits
 - `current.1` `update.sh` will start `current.1` `run.sh` because `restartInteractiveRunner=true`
`run.sh` exited, `current.1` `run.sh` running

### This is how it would migrate:
`current` updating to `patched`:
 - `SelfUpdate` is called with `restartInteractiveRunner=true` (still running `current` binaries)
 - `current` `run.sh` exits
 - `patched` (and unmodified) `update.sh` will start `patched` `run.sh` because `restartInteractiveRunner=true`

`run.sh` exited, `patched` `run.sh` running

### This is how it would work onward:
`patched` updating to `patched.1`:
 - `SelfUpdate` is called with `restartInteractiveRunner=false` (or clean up the arg)
 - `patched` `run.sh` **never** exits
 - `patched.1` `update.sh` will not start anything

`run.sh` never exits, is never replaced

There is a problem with this:

> `run.sh` never exits, is never replaced

This is fine for most updates (is it?), but how would we deliver changes to `run.sh`?

We could perhaps restart it only if the update brings changes to `run.sh`.

Or with each update, have the old `run.sh` start `run.2.284.1.sh`, then `run.2.284.2.sh`, never stopping the original?

Reply:
I think this will work because both `update.cmd` and `update.sh` will restart the runner via `run.sh/cmd` which is going to be the new version you added.